Design.md

```diff
@@ -0,0 +1,107 @@
+
+Meets all requirements (except perhaps around the last stripe size of an ECN chunk). Split is required to contact the same hyperdrives for the object. This has several benefits: only selecting once, object located on only a handful of hyperdrives => less machines to contact = less opportunities to fail or hit a straggler. The way split is handled also enables us not to have a manifest, worries about its size, location or freshness.
+
+Example key:
+1/ Small key: storing s3://fakebucket/obj1 with RS2+1 onto hd1 hd2 and hd3
+Main key: 1#1#1,0#RS,2,1#obj1#314159#hd1#hd3#hd2
+Hyperdrive keys:
+1. obj1-314159-0-d-1
+2. obj1-314159-0-d-2
+3. obj1-314159-0-c-3
+
+2/ Splitted key: storing s3://fakebucket/Large1 with RS2+1 onto hd1 hd2 and hd3
+Key size: 64000, split_size: 49000, 2 parts on same hyperdrive (no conflict!)
+
+Main key: 1#1#2,49000#RS,2,1#Large1#42#hd3#hd2#hd3
+Hyperdrive keys:
+* On hd1: none
+* On hd2
+  1. Large1-42-0-d-2
+  2. Large1-42-49000-d-2
+* On hd3
+  1. Large1-42-0-d-1
+  2. Large1-42-0-c-3
+  3. Large1-42-49000-d-1
+  4. Large1-42-49000-c-3
+
+## Error handling
+
+### On DELETE
+
+Failure to delete a fragment ( != 20X, 404) should not fail the overall operation but must log an appropriate entry into designated Kafka topic, to be cleaned later on by HdRepair.
+
+Topic: orphans
+Content: JSON messages
+Example JSON:
+```json
+{
+    "objectKey": ... # S3 object key
+    "rawKey": .... # associated hdclient metadata
+    "orphans": [
+        hyperdrive_key_missing_fragment1,
+        hyperdrive_key_missing_fragment2,
+        hyperdrive_key_missing_fragment3,
+        ...
+    ]
+}
+```
+
+### On PUT
+
+Failure ot pUT any fragment (erasure coded, replicated or split) should fail the overall operation. All fragments already PUT pending must be cleaned afterwards. Cleaning is handled in the same way as orphans created on deletion (refer to On DELETE section).
+
+Special case of timeout/connection close: a timeout on a PUTing a fragment leaves no clear signal. Is it stored? Queued but not yet written? We could either:
+1. Consider it as not OK and fail the overall PUT
+2. Consider it OK BUT we must log a Kafka entry requesting to check its real status asap (perhaps logging it as a repair action)
+
+Note: we may accept such unsure framgnet trick IFF we have enough framgents already stored. for replication, we must be sure of at least 1 (or 2?). For Reed-Solomon k+m we need at least k real OK (or k+1?).
+
+### On GET
+
+If we are able to provide the caller with the data, we must do so. For replication it equals being able to contact and read data from a single hyperdrive. For erasure coding we must repair data online if we can. In any case, a detected error (404, correupted or else) must be logged and a corresmonding Kafka entry must be persisted, asking hdrepair to check and do its magic if need be. Only exception is failure to contact a hyperdrive.
```

Comment:
What would kafka be able to do in case of 404 ? Reconstruct from remote HyperDrives ?

Does this mean that we consider that a GET should never be queried  for a non-existing key on the HyperDrives ? Can CloudServer guarantee that ?

Reply:
Objects are erasure coded, so we can reconstruct them live (potentially). The idea is to notify a real repair process (through Kafka).